### PR TITLE
Add desktop game template

### PR DESCRIPTION
**Pull request by iiSaLMaN**
- Provides a game template for the desktop platform alongside a test browser. (other platforms are upcoming separately as discussed)

### CODE REVIEW

**Comment by iiSaLMaN** [comment]
@peppy `.slnf` file added and applied project improvements.

File: template-game/TemplateGame.Game.props [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<!-- Contains required properties for osu!framework projects. -->
+<Project>
+  <PropertyGroup Label="C#">
+    <LangVersion>7.2</LangVersion>
```
Comment (huoyaoyuan) [inline template-game/TemplateGame.Game.props]: No reason for keeping an old version here.
- Use the latest known version when template creating, that's `8.0`.
- Use the latest widely-supported version when template creating, that's `7.3`.
- Keep `default` here.
- Don't set it. All .NET Core built-in templates don't set language version.

File: template-game/TemplateGame.Desktop/Program.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+using System;
+using osu.Framework.Platform;
+using osu.Framework;
+using TemplateGame.Game;
+
+namespace TemplateGame.Desktop
+{
+    public static class Program
+    {
+        [STAThread]
```
Comment (huoyaoyuan) [inline template-game/TemplateGame.Desktop/Program.cs]: Is this still required in .NET Core world?
I remember it's a requirement of WinForms.

File: template-game/TemplateGame.Game.Tests/TemplateGame.Game.Tests.csproj [hunk 3]
```diff
@@ -0,0 +1,19 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <Import Project="..\TemplateGame.Game.props" />
```
Comment (huoyaoyuan) [inline template-game/TemplateGame.Game.Tests/TemplateGame.Game.Tests.csproj]: I'd prefer `Directory.Build.props` that doesn't need explicit import.

File: template-game/TemplateGame.Desktop/TemplateGame.Desktop.csproj [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    <OutputType>WinExe</OutputType>
+    <Description>osu!framework-based template game</Description>
+    <AssemblyName>TemplateGame</AssemblyName>
+    <Title>o!f-based game</Title>
+    <ApplicationIcon>game.ico</ApplicationIcon>
+    <Version>0.0.0</Version>
+    <FileVersion>0.0.0</FileVersion>
+  </PropertyGroup>
+  <PropertyGroup>
+    <StartupObject>TemplateGame.Desktop.Program</StartupObject>
```
Comment (huoyaoyuan) [inline template-game/TemplateGame.Desktop/TemplateGame.Desktop.csproj]: Don't think them necessary for a template.

File: template-game/.template.config/template.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{
+    "$schema": "http://json.schemastore.org/template",
+    "author": "ppy Pty Ltd",
+    "classifications": [
+        "Console"
```
Comment (huoyaoyuan) [inline template-game/.template.config/template.json]: Console? Well, it does builds and runs with console runtime.

File: template-game/TemplateGame.sln.DotSettings [hunk 6]
```diff
@@ -0,0 +1,889 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
Comment (huoyaoyuan) [inline template-game/TemplateGame.sln.DotSettings]: Why including this? Everyone could configure their own code convention.
If special attention required when using the framework, an code analyzer should be shipped via package reference.

File: template-game/TemplateGame.Game.props (hunk 1 above)
Comment (iiSaLMaN) [reply template-game/TemplateGame.Game.props]: I guess not setting it is the best option.

File: template-game/TemplateGame.Desktop/Program.cs (hunk 2 above)
Comment (iiSaLMaN) [reply template-game/TemplateGame.Desktop/Program.cs]: It's still used in `SampleGame.Desktop.Program` though?

Comment (huoyaoyuan) [reply template-game/TemplateGame.Desktop/Program.cs]: Yes, but try removing it and see what happens. Or you can remove the corresponding one in main game.
I'm debugging some packaging issue and SampleGame is not runnable in my current branch.

File: template-game/TemplateGame.sln.DotSettings (hunk 6 above)
Comment (iiSaLMaN) [reply template-game/TemplateGame.sln.DotSettings]: Keep in mind I use [ppy/osu-templates](https://www.github.com/ppy/osu-templates) as a reference for how templates are created here, though I agree about anyone configuring their own code convention, Not quite sure about this...

File: template-game/.template.config/template.json (hunk 5 above)
Comment (iiSaLMaN) [reply template-game/.template.config/template.json]: Same as last comment, I use osu! ruleset templates as a reference.

File: template-game/Directory.Build.props [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<!-- Contains required properties for osu!framework projects. -->
+<Project>
+  <PropertyGroup>
+    <ApplicationManifest>..\app.manifest</ApplicationManifest>
```
Comment (huoyaoyuan) [inline template-game/Directory.Build.props]: Note: **never** use `.` or `..` in shared props file, as it interprets relatively to the actual project file, and will break if you projects are reordered into folders. Use `$(MSBuildThisFileDirectory)` and `$([MSBuild]::NormalizePath( ))` instead. This should be fixed for the whole repository, not only this PR. 
See https://docs.microsoft.com/en-us/visualstudio/msbuild/property-functions?view=vs-2019

File: template-game/TemplateGame.sln.DotSettings (hunk 6 above)
Comment (iiSaLMaN) [reply template-game/TemplateGame.sln.DotSettings]: While thinking about this for a bit, I suppose `TemplateGame.sln.DotSettings` should exist in case we add more stuff into the template for consistent code-style with the other projects. Should I safely resolve this conversation?

Comment (smoogipoo) [reply template-game/TemplateGame.sln.DotSettings]: As stated, the `DotSettings` should not be included. It's up to the consumer to use whatever code style they want.

Comment (peppy) [reply template-game/TemplateGame.sln.DotSettings]: We can probably make a `strict` template in the future which has pre-populated formatting/analysers (I would want this for personal use), but let's remove for now.

**Comment by peppy** [comment]
This should likely be added to the deploy process (to generate a nupkg artifact, even for CI builds).

**Comment by peppy** [comment]
Have you tested this? The nupkg doesn't install any new templates for me.

**Comment by iiSaLMaN** [comment]
> Have you tested this? The nupkg doesn't install any new templates for me.

Though this would happen. I researched for like days but I couldn't find a way to pack a template (may be that my research is bad), need help...

**Comment by iiSaLMaN** [comment]
@peppy Should be good to go now!

**Comment by bdach** [comment]
Works with `dotnet new -i` now, but there is a `.sln` file which lands in a slightly wrong place (in the `content/` subdir) and thus that solution fails to load. Also there is no `template.Game.Desktop` project in the resulting structure, its files seem to loosely reside in `content/`.

**Comment by bdach** [comment]
I think I sort of got it to do what's expected. 3f66732 can be reverted, and a root `.csproj` file in the `template-game` directory seems to be needed; see bdach/osu-framework@c96e57b for what I mean.

Basically what you want to do is keep the existing solution as-is, and then add a *separate* new single project that includes all of the files you want to have in the template, *as content*. It's also important to include `.template.config` in the nupkg, else `dotnet new -i` will silently fail (which was determined by anger-laden trial and error).

To deploy, run `dotnet pack osu.Framework.Template.csproj`. The csproj name has to be explicitly supplied to avoid confusion with the solution.

What I came up with works fine with `dotnet new` but doesn't play nice with Rider (and probably VS2019 too, but haven't tested), since Rider wants to create its own `.sln` file.

Not sure if we want to remove all the metadata from `TemplateGame.Game` now then to be filled by the template consumer.

File: template-game/osu.Framework.Template.csproj [hunk 8]
```diff
@@ -0,0 +1,27 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+  <PropertyGroup Label="NuGet">
+    <Title>osu!framework game template</Title>
+    <Product>osu!framework game template</Product>
+    <Authors>ppy Pty Ltd</Authors>
+    <Company>ppy Pty Ltd</Company>
+    <Description>Template to use when creating a game powered by osu!framework.</Description>
```
Comment (smoogipoo) [inline template-game/osu.Framework.Template.csproj]:
```suggestion
    <Description>Create a game powered by osu!framework.</Description>
```

File: template-game/osu.Framework.Template.csproj [hunk 9]
```diff
@@ -0,0 +1,27 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+  <PropertyGroup Label="NuGet">
+    <Title>osu!framework game template</Title>
+    <Product>osu!framework game template</Product>
+    <Authors>ppy Pty Ltd</Authors>
+    <Company>ppy Pty Ltd</Company>
```
Comment (smoogipoo) [inline template-game/osu.Framework.Template.csproj]: Unnecessary, included in the parenting `Directory.Build.props`.

```suggestion
```

File: template-game/osu.Framework.Template.csproj [hunk 10]
```diff
@@ -0,0 +1,27 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+  <PropertyGroup Label="NuGet">
+    <Title>osu!framework game template</Title>
+    <Product>osu!framework game template</Product>
+    <Authors>ppy Pty Ltd</Authors>
```
Comment (smoogipoo) [inline template-game/osu.Framework.Template.csproj]: Unnecessary, included in the parenting `Directory.Build.props`.

```suggestion
```

File: template-game/osu.Framework.Template.csproj [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    <Company>ppy Pty Ltd</Company>
+    <Description>Template to use when creating a game powered by osu!framework.</Description>
+    <PackageId>ppy.osu.Framework.Templates</PackageId>
+    <PackageType>Template</PackageType>
+    <PackageReleaseNotes>Automated release.</PackageReleaseNotes>
```
Comment (smoogipoo) [inline template-game/osu.Framework.Template.csproj]: Unnecessary, included in the parenting `Directory.Build.props`.

```suggestion
```

File: template-game/osu.Framework.Template.csproj [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    <PackageTags>osu game template framework</PackageTags>
+    <Copyright>Copyright (c) 2019 ppy Pty Ltd</Copyright>
```
Comment (smoogipoo) [inline template-game/osu.Framework.Template.csproj]: Unnecessary, included in the parenting `Directory.Build.props`.

```suggestion
```

File: template-game/osu.Framework.Template.csproj [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    <PackageLicenseExpression>MIT</PackageLicenseExpression>
```
Comment (smoogipoo) [inline template-game/osu.Framework.Template.csproj]: Unnecessary, included in the parenting `Directory.Build.props`.

```suggestion
```

File: template-game/osu.Framework.Template.csproj [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    <PackageProjectUrl>https://github.com/ppy/osu-framework/tree/master/template-game</PackageProjectUrl>
+    <RepositoryUrl>https://github.com/ppy/osu-framework/tree/master/template-game</RepositoryUrl>
+	<IncludeContentInPack>true</IncludeContentInPack>
+	<IncludeBuildOutput>false</IncludeBuildOutput>
+	<ContentTargetFolders>templates</ContentTargetFolders>
```
Comment (smoogipoo) [inline template-game/osu.Framework.Template.csproj]: Wrong indentation

File: template-game/osu.Framework.Template.csproj [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+  </PropertyGroup>
+  <ItemGroup>
+    <Content Include=".template.config\*;app.manifest;Directory.Build.props;TemplateGame*.*;TemplateGame*\**\*" Exclude="**\bin\**;**\obj\**" />
+    <Compile Remove="**\*" />
```
Comment (smoogipoo) [inline template-game/osu.Framework.Template.csproj]: Are these required? Surely they can be removed...

File: template-game/TemplateGame.Game/TemplateGameGame.cs [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+using osu.Framework.Allocation;
+
+namespace TemplateGame.Game
+{
+    public class TemplateGameGame : osu.Framework.Game
+    {
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            // todo: add game components and visuals.
```
Comment (smoogipoo) [inline template-game/TemplateGame.Game/TemplateGameGame.cs]:
```suggestion
            // Add your game components here
```

File: template-game/TemplateGame.Game.Tests/Visual/TestSceneTemplateGameGame.cs [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+        [BackgroundDependencyLoader]
+        private void load(GameHost host)
+        {
+            game = new TemplateGameGame();
+            game.SetHost(host);
+
+            Add(game);
+        }
+
+        // todo: add visual tests to ensure correct behaviour of your game: https://github.com/ppy/osu-framework/wiki/Development-and-Testing
```
Comment (smoogipoo) [inline template-game/TemplateGame.Game.Tests/Visual/TestSceneTemplateGameGame.cs]:
```suggestion
        // Add visual tests to ensure correct behaviour of your game: https://github.com/ppy/osu-framework/wiki/Development-and-Testing
```

File: template-game/osu.Framework.Template.csproj (hunk 15 above)
Comment (bdach) [reply template-game/osu.Framework.Template.csproj]: The content inclusion clause should probably there - it's there so that some things that don't need to don't get packaged when running `dotnet pack` (in particular, this fake csproj itself, or `.idea` folders, binaries, etc.). The compile line could probably be removed though, so long that this csproj doesn't end up trying to build the projects.

**Comment by smoogipoo** [comment]
I don't think we can use `o!f-game`... https://streamable.com/b3s74

**Comment by smoogipoo** [comment]
Should add in the vscode task + launch configurations.

**Comment by peppy** [comment]
I already mentioned this in discord and @bdach mentioned [above](https://github.com/ppy/osu-framework/pull/2908#issuecomment-549147461). It should be `osu-framework-game` or similar.

**Comment by smoogipoo** [comment]
.gitignore should be included.

**Comment by peppy** [comment]
As discussed on discord, as `dotnet new gitignore` exists we shouldn't need to ignore the file with our template.

File: template-game/osu.Framework.Template.csproj [hunk 18]
```diff
@@ -0,0 +1,21 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (huoyaoyuan) [inline template-game/osu.Framework.Template.csproj]: Better to change the template to netstandard2.1 as early as possible - See ppy/osu#6837 .

**Comment by peppy** [comment]
@iiSaLMaN packing still doesn't work because of missing nuget restore, as can be seen by failing CI. Looks like it broke in https://github.com/ppy/osu-framework/pull/2908/commits/6d7880a97e0741ff556ca0c41c3895a2528f8bb4.

File: template-game/TemplateGame.Desktop/TemplateGame.Desktop.csproj [hunk 19]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup Label="Project">
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <OutputType>WinExe</OutputType>
+    <Description>osu!framework-based template game</Description>
```
Comment (peppy) [inline template-game/TemplateGame.Desktop/TemplateGame.Desktop.csproj]: why are these here?

File: template-game/.template.config/template.json [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+    ],
+    "name": "osu!framework-based game",
+    "identity": "ppy.osu.Framework.Templates",
```
Comment (peppy) [inline template-game/.template.config/template.json]: these don't match the updated titles in the main csproj

File: template-game/TemplateGame.Game/TemplateGame.Game.csproj [hunk 21]
```diff
@@ -0,0 +1,8 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="ppy.osu.Framework" Version="2019.1121.0" />
```
Comment (peppy) [inline template-game/TemplateGame.Game/TemplateGame.Game.csproj]: needs to be brought up-to-date.

File: template-game/osu.Framework.Template.csproj (hunk 18 above)
Comment (peppy) [reply template-game/osu.Framework.Template.csproj]: hasn't been updated here.

File: template-game/TemplateGame.Desktop/TemplateGame.Desktop.csproj (hunk 19 above)
Comment (iiSaLMaN) [reply template-game/TemplateGame.Desktop/TemplateGame.Desktop.csproj]: Following `osu!` projects setup, will remove.

**Comment by peppy** [comment]
Have you tested this latest version? It doesn't look to work.

**Comment by iiSaLMaN** [comment]
@peppy Tested it before push and right now, working fine?

Also, you do know it starts as black screen, right? maybe I should add a coloured box or a text saying "Template".

**Comment by peppy** [comment]
I'm not seeing the template show up in my list after a `dotnet new --install`. Have you tested from scratch?

**Comment by iiSaLMaN** [comment]
Fully-tested from scratch, installed template and it showed up for me and working fine, 
I downloaded the template nupkg from d08f86b's [AppVeyor generated template artificat](https://ci.appveyor.com/api/buildjobs/sh9nmwagusq0s4qf/artifacts/ppy.osu.Framework.Templates.0.0.11726.nupkg), then installed it by `dotnet new -i .\ppy.osu.Framework.Templates.0.0.11726.nupkg`.

**Comment by peppy** [comment]
Aha, seems to work on another try. I think I had a stale package in my downloads folder which tab completed.

Yeah, I'd recommend adding a spinning cube or something just so it's apparent things are working.

File: template-game/TemplateGame.Game/TemplateGameGame.cs [hunk 22]
```diff
@@ -2,15 +2,32 @@
...
+
+            Child = box = new Box
+            {
+                Anchor = Anchor.Centre,
+                Origin = Anchor.Centre,
+                Colour = Color4.Orange,
+                Size = new Vector2(200),
+            };
+
+            box.Loop(b => b.RotateTo(0).RotateTo(360, 2500));
```
Comment (peppy) [inline template-game/TemplateGame.Game/TemplateGameGame.cs]: move to `LoadComplete` (highly not recommended to do transforms in BDL - this only works due to looping, but is not a good starting point for new framework users).
